common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,1063 @@
+  uint64_t match_bits = vreinterpret_u64_u8(metadata_vec)[0];
+  // The matched range is likely to be tested for zero by the caller, and that
+  // test can often be folded into masking the bits with `MSBs` when we do that
+  // mask in the scalar domain rather than the SIMD domain. So we do the mask
+  // here rather than above prior to extracting the match bits.
+  result = MatchRange(match_bits & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  // We arrange the byte vector for present bytes so that we can directly
+  // extract it as our bit-encoded match.
+  result = MatchRange(_mm_movemask_epi8(metadata_vec));
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchEmpty() const -> MatchIndex {
+  MatchIndex result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto cmp_vec = vceqz_u8(metadata_vec);
+  uint64_t metadata_bits = vreinterpret_u64_u8(cmp_vec)[0];
+  // The matched range is likely to be tested for zero by the caller, and that
+  // test can often be folded into masking the bits with `MSBs` when we do that
+  // mask in the scalar domain rather than the SIMD domain. So we do the mask
+  // here rather than above prior to extracting the match bits.
+  result = MatchIndex(metadata_bits & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = static_cast<MatchIndex>(X86SIMDMatch(Empty));
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchDeleted() const -> MatchIndex {
+  MatchIndex result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto cmp_vec = vceq_u8(metadata_vec, vdup_n_u8(Deleted));
+  uint64_t match_bits = vreinterpret_u64_u8(cmp_vec)[0];
+  // The matched range is likely to be tested for zero by the caller, and that
+  // test can often be folded into masking the bits with `MSBs` when we do that
+  // mask in the scalar domain rather than the SIMD domain. So we do the mask
+  // here rather than above prior to extracting the match bits.
+  result = MatchIndex(match_bits & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = static_cast<MatchIndex>(X86SIMDMatch(Deleted));
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDCompareEqual(MetadataGroup lhs,
+                                            MetadataGroup rhs) -> bool {
+#if CARBON_NEON_SIMD_SUPPORT
+  return vreinterpret_u64_u8(vceq_u8(lhs.metadata_vec, rhs.metadata_vec))[0] ==
+         static_cast<uint64_t>(-1LL);
+#elif CARBON_X86_SIMD_SUPPORT
+  // Different x86 SIMD extensions provide different comparison functionality.
+#if __SSE4_2__
+  return _mm_testc_si128(_mm_cmpeq_epi8(lhs.metadata_vec, rhs.metadata_vec),
+                         _mm_set1_epi8(0xff)) == 1;
+#else
+  return _mm_movemask_epi8(_mm_cmpeq_epi8(lhs.metadata_vec,
+                                          rhs.metadata_vec)) == 0x0000'ffffU;
```

Comment:
Seems like there are a lot of ways this could be done -- any insight into why these particular instructions are used?